--- a/drivers/net/phy/rtl8367b.c
+++ b/drivers/net/phy/rtl8367b.c
@@ -1517,6 +1517,12 @@ static int rtl8367b_detect(struct rtl836
 			smi->rtl8367b_chip = RTL8367B_CHIP_RTL8367S;
 		}
 		break;
+	case 0x0030:
+		if (chip_num == 0x6367) {
+			chip_name = "8367N";
+			smi->rtl8367b_chip = RTL8367B_CHIP_RTL8367S;
+		}
+		break;
 	case 0x1000:
 		chip_name = "8367RB";
 		smi->rtl8367b_chip = RTL8367B_CHIP_RTL8367RB;
